File: go/reconstructBST/reconstructBST.go
package main


import (
	"fmt"
	"math"
	"reflect"
)

type BST struct {
	Value int

	Left  *BST
	Right *BST
}

func (tree *BST) Insert(value int) {
	for tree != nil {
		if value < tree.Value {
			if tree.Left == nil {
				tree.Left = &BST{Value: value}
				break
			} else {
				tree = tree.Left
			}
		} else {
			if tree.Right == nil {
				tree.Right = &BST{Value: value}
				break
			} else {
				tree = tree.Right
			}
		}
	}
}

func (tree *BST) PreOrder(array []int) []int {
	if tree == nil {
		return array
	}

	array = append(array, tree.Value)
	array = tree.Left.PreOrder(array)
	array = tree.Right.PreOrder(array)

	return array
}


func BuildFromPreOrder(array []int) *BST{
	bst := &BST{Value: array[0]}
	for _, v := range array[1:] {bst.Insert(v)}
	return bst
}


func BuildFromPreOrder2(array []int) *BST{
	return RecursiveAddition(math.MinInt, math.MaxInt, array, []int{0})
}


func RecursiveAddition(lowerBound, upperBound int, array, rootidx []int) *BST{
	if rootidx[0] == len(array) {return nil}

	rootValue := array[rootidx[0]]
	if rootValue < lowerBound || rootValue >= upperBound {return nil}

	rootidx[0] += 1
	leftsubtree := RecursiveAddition(lowerBound, rootValue, array, rootidx)
	rightsubtree := RecursiveAddition(rootValue, upperBound, array, rootidx)

	return &BST{
		Value: rootValue,
		Left: leftsubtree,
		Right: rightsubtree,
	}

}


func main() {
	array := []int{10, 4, 2, 1, 5, 17, 19, 18}

	bst := BuildFromPreOrder(array)
	bst2 := BuildFromPreOrder2(array)
	fmt.Println(reflect.DeepEqual(bst.PreOrder([]int{}), array))
	fmt.Println(reflect.DeepEqual(bst2.PreOrder([]int{}), array))
}

File: python/reconstructBST.py
class BST:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right
    
    
    def Insert(self, value):
        while self != None:
            if value < self.value:
                if self.left == None:
                    self.left = BST(value)
                    break
                else:
                    self = self.left
            else:
                if self.right == None:
                    self.right = BST(value)
                    break
                else:
                    self = self.right
                    
    
    def PreOrder(self, array):
        array.append(self.value)
        if self.left != None:
            self.left.PreOrder(array)
        if self.right != None:
            self.right.PreOrder(array)
        return array
    
    
    
preOrderTraversal = [10, 4, 2, 1, 5, 17, 19, 18]


# Time -> 0(NlogN) Space -> 0(h)
def buildFromPreOrder(array):
    bst = BST(array[0])
    for val in array[1:]:
        bst.Insert(val)
        
    return bst


# Time -> 0(N) Space -> 0(h)    h is the height of the tree 
def buildFromPreOrder2(array):
    return RecursiveAddition(float("-inf"), float("inf"), array, [0])
    

def RecursiveAddition(lowerBound, upperBound, array, rootidx):
    if rootidx[0] == len(array):
        return None
    
    rootvalue = array[rootidx[0]]
    if rootvalue < lowerBound or rootvalue >= upperBound:
        return None
    
    rootidx[0] += 1
    leftsubtree = RecursiveAddition(lowerBound, rootvalue, array, rootidx)
    rightsubtree = RecursiveAddition(rootvalue, upperBound, array, rootidx)
    
    return BST(rootvalue, leftsubtree, rightsubtree)
        

bst = buildFromPreOrder(preOrderTraversal)
bst2 = buildFromPreOrder2(preOrderTraversal)
print(bst2.PreOrder([]) == preOrderTraversal)
print(bst.PreOrder([]) == preOrderTraversal)
